plugin_tests/docker_test.py

```diff
@@ -0,0 +1,280 @@
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+###############################################################################
+
+# This is to serve as an example for how to create a server-side test in a
+# girder plugin, it is not meant to be useful.
+
+from tests import base
+
+from girder import events
+import threading
+import six
+import types
+import json
+# boiler plate to start and stop the server
+TIMEOUT = 180
+
+
+def setUpModule():
+    base.enabledPlugins.append('HistomicsTK')
+    base.startServer()
+    global JobStatus
+    from girder.plugins.jobs.constants import JobStatus
+    global DockerImage
+    from girder.plugins.HistomicsTK.models.docker_image import DockerImage
+
+
+def tearDownModule():
+    base.stopServer()
+
+# TODO when endpoint information is added to the get endpoint
+# TODO add tests for querying cli xml and running clis
+
+
+class HistomicsTKExampleTest(base.TestCase):
+    def setUp(self):
+        # adding and removing docker images and using generated rest endpoints
+        # requires admin access
+        base.TestCase.setUp(self)
+        admin = {
+            'email': 'admin@email.com',
+            'login': 'adminlogin',
+            'firstName': 'Admin',
+            'lastName': 'Last',
+            'password': 'adminpassword',
+            'admin': True
+        }
+        self.admin = self.model('user').createUser(**admin)
+
+    def testBadImageAdd(self):
+        # add a bad image
+        img_name = 'null/null:null'
+        self.noImages()
+        self.addImage(img_name, JobStatus.ERROR)
+        self.noImages()
+
+    def testDockerAdd(self):
+        # try to cache a good image to the mongo database
+        img_name = "dsarchive/histomicstk:v0.1.3"
+        self.noImages()
+        self.addImage(img_name, JobStatus.SUCCESS)
+        self.imageIsLoaded(img_name, True)
+
+    def testDockerDelete(self):
+        # just delete the meta data in the mongo database
+        # dont attempt to delete the docker image
+        img_name = "dsarchive/histomicstk:v0.1.3"
+        self.noImages()
+        self.addImage(img_name, JobStatus.SUCCESS)
+        self.imageIsLoaded(img_name, True)
+        self.deleteImage(img_name, True, False)
+        self.imageIsLoaded(img_name, exists=False)
+        self.noImages()
+
+    def testDockerDeleteFull(self):
+        # attempt to delete docker image metadata and the image off the local
+        # machine
+        img_name = "dsarchive/histomicstk:v0.1.3"
+        self.noImages()
+        self.addImage(img_name, JobStatus.SUCCESS)
+        self.imageIsLoaded(img_name, True)
+        self.deleteImage(img_name, True, True, JobStatus.SUCCESS)
+        try:
+            from docker import Client
+            docker_client = Client(base_url='unix://var/run/docker.sock')
+
+        except Exception as err:
+            self.fail('could not create the docker client '+err.__str__())
```

Comment:
You may want to move the loading of `docker_client` into the setup function so multiple test functions could use it if they want.